### PR TITLE
Unbranded Multi card fields POC

### PR DESCRIPTION
**Pull request by leogedler**
### Description

POC: Multi card field rendering and styling.

This is part of the effort relate to the new PayPal card field component.

### Why are we making these changes? Include references to any related Jira tasks or GitHub Issues

https://engineering.paypalcorp.com/jira/browse/DTNOR-222
https://engineering.paypalcorp.com/jira/browse/DTNOR-225

### Screenshots (if applicable)

image

image

image

image

Related PR

https://github.com/paypal/paypal-sdk-samples/pull/2

❤️  Thank you!

### CODE REVIEW

File: src/card/lib/card-utils.test.js [hunk 1]
```diff
@@ -106,5 +106,38 @@ describe('card utils', () => {
             expect(spy).toBeCalledWith('');
         });
     });
+
+    describe('maskDate', () => {
```
Comment (crookedneighbor) [inline src/card/lib/card-utils.test.js:110]: Naming nit pick, but this looks more like a `formatDate` function than a `maskDate` function.

File: src/card/components/CardNumber.jsx [hunk 2]
```diff
@@ -105,7 +105,10 @@ export function CardNumber(
             endCursorPosition = cursorEnd;
         }
         
-        if (maskedInputValue.length !== maskedValue.length && maskedValue.length === selectionStart + 1) {
+        if (contentPasted) {
+            startCursorPosition = maskedValue.length;
+            endCursorPosition  = maskedValue.length;
```
[the remark this answers is not in the document]
Comment (crookedneighbor) [reply src/card/components/CardNumber.jsx:110]: I took out the cursor movement in the masking on my masking PR, and it worked well enough after that.

File: src/card/lib/card-utils.test.js (hunk 1 above)
Comment (leogedler) [reply src/card/lib/card-utils.test.js:110]: Got it, I'll change it.

File: src/card/components/CardNumber.jsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (leogedler) [reply src/card/components/CardNumber.jsx:110]: Interesting, the original problem was reported by @nbierdeman , when the user pastes a card number the cursor was place somewhere in the middle of the card number.

This is why I'd like to have those changes from the masking improvement here in this PR, that way we can re-check if the cursos problem is also gone for the multi card fields (I bet they will if it work for the single card fields).

File: src/card/components/fields.jsx [hunk 3]
```diff
@@ -138,39 +139,45 @@ export function CardField({ cspNonce, onChange, styleObject = {}, placeholder =
...
-    const [ number, setNumber ] = useState('');
-    const [ numberValidity, setNumberValidity ] = useState(true);
+export function CardNumberField({ cspNonce, onChange, styleObject = {}, placeholder = {}, autoFocusRef } : CardNumberFieldProps) : mixed {
+    const [ number, setNumber ] : [ string, (string) => string ] = useState('');
+    const [ numberValidity, setNumberValidity ] : [ FieldValidity, (FieldValidity) => FieldValidity ] = useState(initFieldValidity);
     const [ generalStyle, inputStyle ] = getStyles(styleObject);
+    const numberRef = useRef();
 
     const composedStyles = { ...{ input: DEFAULT_INPUT_STYLE },  ...generalStyle };
+    const { isValid, isPossibleValid } = numberValidity;
```
Comment (crookedneighbor) [inline src/card/components/fields.jsx:170]: Over on Braintree's SDK, we use `isPotentiallyValid` over `isPossibleValid`, which I think reads a little better. Feel free to use whatever makes sense to y'all though.

**Review by crookedneighbor — COMMENTED** [review]
Maybe not for this PR, but did y'all want to consider using the input formatting library that Braintree developed? 

Repo: https://github.com/braintree/restricted-input
Demo: https://braintree.github.io/restricted-input/

**Comment by leogedler** [comment]
@crookedneighbor maybe I'm missing something in the formatting library code but I don't see a direct way to implement it on React/Preact components. It appears to be just vanilla JS.

I'm not against implementing new libraries but initially when we started working on this feature the signs that I got were to try to make it simple and without too much external dependencies.

For example when I started working on the styling my initial approach was trying to implement https://emotion.sh/docs/introduction to simplify the handling of the style composition but then @westeezy told me that we really don't need it to accomplish the goal. In the end @emotion wasn't necessary after all.

**Comment by crookedneighbor** [comment]
Yeah, in Braintree, we opted to do everything platform agnostic, so it's all vanilla JS. 

Feel free to ignore my suggestion, just wanted to point out that we have an already built input formatting library that is not tied to credit card specifically.

**Comment by nbierdeman** [comment]
@leogedler, I noticed a few positions in the card number field where deleting a character results in the cursor jumping unexpectedly. For example, here I'm attempting to go from `1244` to `1234`:

ezgif com-gif-maker

**Comment by leogedler** [comment]
@nbierdeman checking.

**Comment by leogedler** [comment]
@nbierdeman I have pushed an improvement for the jumping cursor while erasing.

**Comment by nbierdeman** [comment]
@leogedler, is multi card fields supposed to support 13-19 digit Visa V Pay numbers? If so, I'm getting an error using the bottom card number from [Test Payflow Transactions](https://developer.paypal.com/docs/payflow/payflow-pro/payflow-pro-testing/):

Screen Shot 2021-11-08 at 2 02 24 PM

**Comment by leogedler** [comment]
@nbierdeman interesting edge case.

For validation we are using the Braintree library [card-validator](https://github.com/braintree/card-validator) but apparently this library is not handling this case.

I think we could add this particular sequence as a custom card to the validation logic.

FYI @crookedneighbor

**Comment by crookedneighbor** [comment]
Re: 13 digit visas, see https://github.com/braintree/credit-card-type/issues/64#issuecomment-363655182

> 13 digit visas are no longer in circulation, so it does not make sense to include them in this module.

Does PayPal require supporting this since it _shouldn't_ exist anymore?

**Comment by nbierdeman** [comment]
> Does PayPal require supporting this since it shouldn't exist anymore?

Good question. I'll follow up with Rahul tomorrow.

**Comment by nbierdeman** [comment]
@leogedler, I'm curious why masking isn't applied to an invalid card number. My thinking is if I enter my credit card number correctly, except for one digit, then move on to the next field, my mostly correct card number is visible on the screen until I review what I entered. I was wondering if it'd make sense to apply masking and styling (e.g. red) to invalid card numbers, once someone has moved away from the field:

Screen Shot 2021-11-08 at 4 20 10 PM

**Comment by leogedler** [comment]
@nbierdeman I get your point, this is working the same way as the checkout UI. (See attachment).

Maybe we could discuss this with the UX designer.

image

File: src/card/components/fields.jsx [hunk 4]
```diff
@@ -69,14 +72,40 @@ export function CardField({ cspNonce, onChange, styleObject = {}, placeholder =
     useEffect(() => {
         autoFocusRef(numberRef);
     }, []);
+
+    useEffect(() => {
+        const { field, errors } = gqlErrorsObject;
+
+        if (field === CARD_FIELD_TYPE.NUMBER) {
+            const hasQQLErros = errors.length > 0;
```
Comment (nbierdeman) [inline src/card/components/fields.jsx:80]: Is this supposed to be this?
```suggestion
            const hasGQLErrors = errors.length > 0;
```

File: src/card/components/fields.jsx [hunk 5]
```diff
@@ -69,14 +72,40 @@ export function CardField({ cspNonce, onChange, styleObject = {}, placeholder =
...
+            }
+        }
+
+        if (field === CARD_FIELD_TYPE.CVV) {
+            const hasQQLErros = errors.length > 0;
+            if (hasQQLErros) {
+                setCvvValidity({ isPossibleValid: false, isValid: false });
+            }
+        }
+
```
Comment (nbierdeman) [inline src/card/components/fields.jsx:99]:
```suggestion

```

File: src/card/components/fields.jsx (hunk 4 above)
Comment (leogedler) [reply src/card/components/fields.jsx]: Thanks! I have applied a fix for this.

File: test/client/card-fields.js [hunk 6]
```diff
@@ -14,342 +14,960 @@ import {
...
+        it('should render multi card fields with createOrder and call onApprove callback', async () => {
+
+            return await wrapPromise(async ({ expect }) => {
+
+                const orderID = generateOrderID();
+                const payerID = 'AAABBBCCC';
+
+                window.xprops.type = CARD_FIELD_TYPE.NUMBER;
+                window.xprops.type = CARD_FIELD_TYPE.EXPIRY;
+                window.xprops.type = CARD_FIELD_TYPE.CVV;
```
Comment (nbierdeman) [inline test/client/card-fields.js:489]: Are these consecutively overwriting the previous `window.xprops.type` values? If so, is that intentional?

**Review by nbierdeman — COMMENTED** [review]
@leogedler, I just finished with my first read-through of your code changes. 🤯  Great work! 
🥇  I'm looking forward to getting this merged soon. 😅

File: test/client/card-fields.js (hunk 6 above)
Comment (leogedler) [reply test/client/card-fields.js:489]: Unintentional and totally unnecessary, thanks for noticing this!.

**Review by nbierdeman — APPROVED** [review]
@leogedler, thanks for addressing my comments so quickly! I went through and resolved the comments that you addressed, but left open the ones that I'm still curious about. Other than that and possibly providing a little more test coverage, this is looking ready to merge. Great work! 🥳

**Review by gregjopa — COMMENTED** [review]
Looks good @leogedler! I left a couple minor comments.

File: src/api/api.js [hunk 7]
```diff
@@ -139,6 +139,11 @@ export function callGraphQL<T>({ name, query, variables = {}, headers = {} } : {
             const message = errors[0].message || JSON.stringify(errors[0]);
 
             getLogger().warn(`graphql_${ name }_error`, { err: message });
+
+            if (returnErrorObject) {
```
[the remark this answers is not in the document]
Comment (leogedler) [reply src/api/api.js:143]: @gregjopa we need to parse the GQL errors for the `approveCardPayment` mutation because then we need to pass it as a code like error to the `onChange` event in the `cardFields` component before throwing the error.

Without the new `returnErrorObject ` option `callGraphQL` always send back an error string that is not containing the full error data.

This is an example of the error coming from `approveCardPayment` mutation that we need to parse with `parseGQLErrors`:

```
{
  "error": {
    "errors": [
      {
        "message": "Status code: 422 {\n  \"name\": \"UNPROCESSABLE_ENTITY\",\n  \"details\": [\n    {\n      \"field\": \"/payment_source/card/number\",\n      \"issue\": \"INVALID_REQUEST\",\n      \"description\": \"The value of a field is either too short or too long\"\n    }\n  ],\n  \"message\": \"The requested action could not be performed, semantically incorrect, or failed business validation.\",\n  \"debug_id\": \"5d1acecc9cd72\",\n  \"links\": [\n    {\n      \"href\": \"https://developer.paypal.com/docs/api/orders/v2/#error-INVALID_REQUEST\",\n      \"rel\": \"information_link\",\n      \"method\": \"GET\"\n    }\n  ]\n}",
        "path": [
          "processPayment"
        ],
        "stack": "Error: Status code: 422 {\n  \"name\": \"UNPROCESSABLE_ENTITY\",\n  \"details\": [\n    {\n      \"field\": \"/payment_source/card/number\",\n      \"issue\": \"INVALID_REQUEST\",\n      \"description\": \"The value of a field is either too short or too long\"\n    }\n  ],\n  \"message\": \"The requested action could not be performed, semantically incorrect, or failed business validation.\",\n  \"debug_id\": \"5d1acecc9cd72\",\n  \"links\": [\n    {\n      \"href\": \"https://developer.paypal.com/docs/api/orders/v2/#error-INVALID_REQUEST\",\n      \"rel\": \"information_link\",\n      \"method\": \"GET\"\n    }\n  ]\n}\n    at AsyncFunction.<anonymous> (/Users/lgedler/Documents/SDK/xobuyernodeserv-shruti/src/graphql/types/Mutation/Mutation.js:4574:23)\n    at processTicksAndRejections (internal/process/task_queues.js:97:5)",
        "data": [
          {
            "code": "UNPROCESSABLE_ENTITY",
            "details": [
              {
                "field": "/payment_source/card/number",
                "issue": "INVALID_REQUEST",
                "description": "The value of a field is either too short or too long"
              }
            ]
          }
        ]
      }
    ],
    "data": {
      "processPayment": null
    },
    "extensions": {
      "correlation_id": "5d1acecc9cd72"
    }
  }
}
```

I have added tests for `parseGQLErrors` in: https://github.com/leogedler/paypal-smart-payment-buttons/blob/multi-card-fields-poc/src/card/lib/card-utils.test.js#L172

File: src/api/order.js [hunk 8]
```diff
@@ -865,7 +865,7 @@ type ApproveCardPaymentOptions = {|
     card : {|
         cardNumber : string,
         expirationDate : string,
-        cvv : string,
```
[the remark this answers is not in the document]
Comment (crookedneighbor) [reply src/api/order.js:876]: Are we going to update this type to show that all but the `number` field is optional?

File: src/card/components/CardCVV.jsx [hunk 9]
```diff
@@ -87,7 +87,7 @@ export function CardCVV(
             onFocus(event);
         }
         if (!isValid) {
-            setInputState({ ...inputState, isPossibleValid: true });
+            setInputState(newState => ({ ...newState, isPossibleValid: true }));
```
[the remark this answers is not in the document]
Comment (crookedneighbor) [reply src/card/components/CardCVV.jsx:90]: Would still love to see this changed from `isPossibleValid` to `isPotentiallyValid`. Just let me know if we've officially come to a decision that `isPossibleValid` is what we want to go with, and I'll shut up about it :)

[the remark this answers is not in the document]
Comment (crookedneighbor) [reply src/card/components/CardCVV.jsx:90]: (Also, admittedly, doing this change is probably better suited for a dedicated PR)

File: src/card/components/page.jsx [hunk 10]
```diff
@@ -35,7 +35,7 @@ function Page({ cspNonce, props } : PageProps) : mixed {
 
     useEffect(() => {
         onChange({
-            valid:    fieldValid,
```
[the remark this answers is not in the document]
Comment (crookedneighbor) [reply src/card/components/page.jsx:40]: agreed. I prefer `isValid`

File: src/card/lib/card-utils.js [hunk 11]
```diff
@@ -144,14 +145,14 @@ export function removeDateMask(date : string) : string {
...
 
-// Mask date
-export function maskDate(date : string, prevMask? : string = '') : string {
+// Format expiry date
+export function formatDate(date : string, prevFormat? : string = '') : string {
     assertString(date);
 
-    if (prevMask && prevMask.indexOf('/') > -1) {
-        const [ month ] = removeSpaces(prevMask).split('/');
+    if (prevFormat && prevFormat.indexOf('/') > -1) {
```
[the remark this answers is not in the document]
Comment (crookedneighbor) [reply src/card/lib/card-utils.js:152]: Yeah, I believe that is correct.

[the remark this answers is not in the document]
Comment (crookedneighbor) [reply src/card/lib/card-utils.js:152]: Got to revert all the IE11 specific changes I made recently :)

File: src/api/order.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (leogedler) [reply src/api/order.js:876]: Sure, I'll do it.

File: src/card/components/page.jsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (leogedler) [reply src/card/components/page.jsx:40]: got it, I'll change it.

File: src/card/lib/card-utils.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (leogedler) [reply src/card/lib/card-utils.js:152]: I'll revert them.

File: src/card/lib/card-utils.test.js [hunk 12]
```diff
@@ -128,5 +135,273 @@ describe('card utils', () => {
...
+                                location:    'body',
+                                issue:       'MISSING_REQUIRED_PARAMETER',
+                                description: 'A required field / parameter is missing.'
+                            }
+                        ]
+                    }
+                ]
+            };
+
+            const { parsedErrors, errors } = parseGQLErrors(gqlError);
```
[the remark this answers is not in the document]
Comment (leogedler) [reply src/card/lib/card-utils.test.js:300]: I'll add it

File: src/card/components/CardCVV.jsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (nbierdeman) [reply src/card/components/CardCVV.jsx:90]: For whatever it's worth (2 cents to be exact), my personal preference is also `isPotentiallyValid`.

[the remark this answers is not in the document]
Comment (leogedler) [reply src/card/components/CardCVV.jsx:90]: Got it, I'll change it.

**Review by crookedneighbor — APPROVED** [review]
